Panic on transaction commit when deleting a large number of keys

Hey there, 

I'm experimenting with bolt and am running into an issue when bulk deleting keys from a database within a transaction. During the commit, bolt hangs, then panics. 

I'm trying to narrow down and extract a piece of code that can reproduce the error, but here are two unique panic dumps I've experienced:

```
panic: tx 11003: page 94877 already freed in tx 11003

goroutine 1 [running]:
runtime.panic(0xc43c0, 0x24c233350)
        /usr/local/go/src/pkg/runtime/panic.c:266 +0xb6
github.com/boltdb/bolt.(*freelist).free(0x21032e040, 0x2afb, 0x2227830000)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/freelist.go:85 +0x3a8
github.com/boltdb/bolt.(*node).spill(0x24cf64150, 0x24cf63620, 0x23d388)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/node.go:307 +0x1d6
github.com/boltdb/bolt.(*node).spill(0x2103989a0, 0x2104f91b0, 0x2210453b50)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/node.go:293 +0xc2
github.com/boltdb/bolt.(*Bucket).spill(0x2104f9180, 0x2103c4c00, 0x1)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/bucket.go:543 +0x182
github.com/boltdb/bolt.(*Bucket).spill(0x210319c48, 0x1cbd09ce, 0x23da40)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/bucket.go:514 +0x7bb
github.com/boltdb/bolt.(*Tx).Commit(0x210319c30, 0x0, 0x244e7)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/tx.go:159 +0x1a9
```

```
panic: assertion failed: put: zero-length old key

goroutine 4 [running]:
runtime.panic(0x1261c0, 0xc2167281d0)
        /usr/local/go/src/pkg/runtime/panic.c:266 +0xb6
github.com/boltdb/bolt._assert(0x0, 0x1bc4d0, 0x18, 0x0, 0x0, ...)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/db.go:637 +0xcb
github.com/boltdb/bolt.(*node).put(0xc2100423f0, 0x0, 0x0, 0x0, 0xaa3dc55, ...)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/node.go:100 +0x158
github.com/boltdb/bolt.(*node).rebalance(0xc216539a80)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/node.go:489 +0x14f8
github.com/boltdb/bolt.(*node).rebalance(0xc218446770)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/node.go:406 +0x35a
github.com/boltdb/bolt.(*Bucket).rebalance(0xc210037d50)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/bucket.go:606 +0x72
github.com/boltdb/bolt.(*Bucket).rebalance(0xc2100621b8)
        /Users/jordan/Development/experiments/gopath/src/github.com/boltdb/bolt/bucket.go:609 +0xde
```

I'm on OS X Mavericks, Go 1.2.1, and on the current boltdb head: 9ffb297 

Any thoughts as to what might be causing this?

--- a/bucket_test.go
+++ b/bucket_test.go
@@ -2,6 +2,7 @@ package bolt
 
 import (
 	"bytes"
+	"encoding/binary"
 	"errors"
 	"fmt"
 	"math/rand"
@@ -183,6 +184,50 @@ func TestBucket_Delete_Large(t *testing.T) {
 			var b = tx.Bucket([]byte("widgets"))
 			for i := 0; i < 100; i++ {
 				assert.Nil(t, b.Get([]byte(strconv.Itoa(i))))
+			}
+			return nil
+		})
+	})
+}
+
+// Ensure that deleting a large set of keys will work correctly.
+// Reported by Jordan Sherer: https://github.com/boltdb/bolt/issues/184
+func TestBucket_Delete_Large2(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
+	}
+
+	withOpenDB(func(db *DB, path string) {
+		k := make([]byte, 16)
+		for i := uint64(0); i < 10000; i++ {
+			err := db.Update(func(tx *Tx) error {
+				b, err := tx.CreateBucketIfNotExists([]byte("0"))
+				if err != nil {
+					t.Fatalf("bucket error: %s", err)
+				}
+
+				for j := uint64(0); j < 1000; j++ {
+					binary.BigEndian.PutUint64(k[:8], i)
+					binary.BigEndian.PutUint64(k[8:], j)
+					if err := b.Put(k, nil); err != nil {
+						t.Fatalf("put error: %s", err)
+					}
+				}
+
+				return nil
+			})
+
+			if err != nil {
+				t.Fatalf("update error: %s", err)
+			}
+		}
+
+		// Delete all of them in one large transaction
+		db.Update(func(tx *Tx) error {
+			b := tx.Bucket([]byte("0"))
+			c := b.Cursor()
+			for k, _ := c.First(); k != nil; k, _ = c.Next() {
+				b.Delete(k)
 			}
 			return nil
 		})

--- a/node.go
+++ b/node.go
@@ -486,7 +486,6 @@ func (n *node) rebalance() {
 		target.inodes = append(target.inodes, n.inodes...)
 		n.parent.del(n.key)
 		n.parent.removeChild(n)
-		n.parent.put(target.key, target.inodes[0].key, nil, target.pgid, 0)
 		delete(n.bucket.nodes, n.pgid)
 		n.free()
 	}

